### PR TITLE
Unit tests for DeviceAccessAuthorizationService

### PR DESCRIPTION
**Pull request by GDLMadushanka**

### CODE REVIEW

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/test/java/org/wso2/carbon/device/mgt/core/authorization/DeviceAccessAuthorizationServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+import org.wso2.carbon.user.api.Permission;
+import org.wso2.carbon.user.core.service.RealmService;
+import org.wso2.carbon.user.core.tenant.JDBCTenantManager;
+import org.wso2.carbon.utils.multitenancy.MultitenantConstants;
+
+import java.io.InputStream;
+import java.util.*;
+
+
+public class DeviceAccessAuthorizationServiceTest extends BaseDeviceManagementTest {
```
Comment (Megala21) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/test/java/org/wso2/carbon/device/mgt/core/authorization/DeviceAccessAuthorizationServiceTest.java]: Shall we add the java doc comment for this test case

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/test/java/org/wso2/carbon/device/mgt/core/authorization/DeviceAccessAuthorizationServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+import org.wso2.carbon.registry.core.jdbc.realm.InMemoryRealmService;
+import org.wso2.carbon.registry.core.service.RegistryService;
+import org.wso2.carbon.user.api.*;
+import org.wso2.carbon.user.api.Permission;
+import org.wso2.carbon.user.core.service.RealmService;
+import org.wso2.carbon.user.core.tenant.JDBCTenantManager;
+import org.wso2.carbon.utils.multitenancy.MultitenantConstants;
+
+import java.io.InputStream;
+import java.util.*;
```
Comment (Megala21) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/test/java/org/wso2/carbon/device/mgt/core/authorization/DeviceAccessAuthorizationServiceTest.java]: It is always good to import only the relevant packages we need, without using wild card character.

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/test/java/org/wso2/carbon/device/mgt/core/authorization/DeviceAccessAuthorizationServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+    private static final String EMAIL = "email";
+    private static final String PASSWORD = "password";
+    private List<DeviceIdentifier> deviceIds = new ArrayList<>();
+    private List<DeviceIdentifier> groupDeviceIds = new ArrayList<>();
+    private Map<String, String> defaultUserClaims;
+
+    @BeforeClass
+    public void init() throws Exception {
+        DeviceConfigurationManager.getInstance().initConfig();
+        log.info("Initializing");
```
Comment (Megala21) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/test/java/org/wso2/carbon/device/mgt/core/authorization/DeviceAccessAuthorizationServiceTest.java]: Can we give a more specific info message ?

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/test/java/org/wso2/carbon/device/mgt/core/authorization/DeviceAccessAuthorizationServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+        realmService.setTenantManager(jdbcTenantManager);
+        RegistryDataHolder.getInstance().setRealmService(realmService);
+        DeviceManagementDataHolder.getInstance().setRealmService(realmService);
+        InputStream is = this.getClass().getClassLoader().getResourceAsStream("carbon-home/repository/conf/registry.xml");
+        RegistryContext context = RegistryContext.getBaseInstance(is, realmService);
+        context.setSetup(true);
+        return context.getEmbeddedRegistryService();
+    }
+
+    @Test
```
Comment (Megala21) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/test/java/org/wso2/carbon/device/mgt/core/authorization/DeviceAccessAuthorizationServiceTest.java]: Please add the description for the test case. Please do it for other test cases as well.

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/test/java/org/wso2/carbon/device/mgt/core/authorization/DeviceAccessAuthorizationServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+    public void userAuthDevIdUserName() throws Exception {
+        for (DeviceIdentifier deviceId : deviceIds) {
+            Assert.assertTrue(deviceAccessAuthorizationService.isUserAuthorized(deviceId, ADMIN_USER),
+                    "Device access authorization for admin user failed");
+        }
+    }
+
+    @Test
+    public void userAuthDevIdUserNameResult() throws Exception {
```
Comment (Megala21) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/test/java/org/wso2/carbon/device/mgt/core/authorization/DeviceAccessAuthorizationServiceTest.java]: Why are we throwing generic exceptions? Please only add the relevant exceptions, that will be thrown

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/test/java/org/wso2/carbon/device/mgt/core/authorization/DeviceAccessAuthorizationServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+    public void userAuthDevId() throws Exception {
+        for (DeviceIdentifier deviceId : deviceIds) {
+            Assert.assertTrue(deviceAccessAuthorizationService.isUserAuthorized(deviceId),
+                    "Authorize user from device identifier failed");
+        }
+    }
+
+    @Test
+    public void userAuthDevIdResult() throws Exception {
+        DeviceAuthorizationResult deviceAuthorizationResult = deviceAccessAuthorizationService.isUserAuthorized(deviceIds);
```
Comment (Megala21) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/test/java/org/wso2/carbon/device/mgt/core/authorization/DeviceAccessAuthorizationServiceTest.java]: Line seems to be having more than 120 characters.

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/test/java/org/wso2/carbon/device/mgt/core/authorization/DeviceAccessAuthorizationServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+        Assert.assertEquals(deviceAuthorizationResult.getAuthorizedDevices().size(), 5,
+                "Expected 5 authorized devices for admin user");
+        Assert.assertEquals(deviceAuthorizationResult.getUnauthorizedDevices().size(), 0,
+                "Expected 0 un-authorized devices for admin user");
+    }
+
+    @Test
+    public void userAuthDevIdPermission() throws DeviceAccessAuthorizationException, UserStoreException, PermissionManagementException {
```
Comment (Megala21) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/test/java/org/wso2/carbon/device/mgt/core/authorization/DeviceAccessAuthorizationServiceTest.java]: 120 character limit seems to be violated.

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/test/java/org/wso2/carbon/device/mgt/core/authorization/DeviceAccessAuthorizationServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+    @Test
+    public void userAuthFalseDevIdPermission() throws DeviceAccessAuthorizationException, UserStoreException,
+            PermissionManagementException {
+        PrivilegedCarbonContext.startTenantFlow();
+        PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantId(MultitenantConstants.SUPER_TENANT_ID, true);
+        PrivilegedCarbonContext.getThreadLocalCarbonContext().setUsername(NON_ADMIN_ALLOWED_USER);
+        Assert.assertFalse(deviceAccessAuthorizationService.isUserAuthorized(deviceIds.get(3), new String[]{NON_ADMIN_PERMISSION}),
+                "Non admin user accessing not allowed device authorized");
+        PrivilegedCarbonContext.endTenantFlow();
+    }
```
Comment (Megala21) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/test/java/org/wso2/carbon/device/mgt/core/authorization/DeviceAccessAuthorizationServiceTest.java]: Why we are starting and ending tenant flow in all the test cases? Shall we move the start tenant flow to @BeforeTest method and shall we end tenant flow in the tear down ?

**Review by Megala21 — CHANGES_REQUESTED** [review]

File: components/device-mgt/org.wso2.carbon.device.mgt.core/pom.xml [hunk 9]
```diff
@@ -359,6 +359,12 @@
 			<groupId>javassist</groupId>
 			<artifactId>javassist</artifactId>
 		</dependency>
+        <dependency>
+            <groupId>org.powermock</groupId>
+            <artifactId>powermock-api-mockito</artifactId>
+            <version>${power.mock.version}</version>
+            <scope>test</scope>
+        </dependency>
```
Comment (Megala21) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/pom.xml]: Shall we move this dependency to parent pom, if it is not already there and remove the version from this pom

**Review by Megala21 — APPROVED** [review]

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/test/java/org/wso2/carbon/device/mgt/core/authorization/DeviceAccessAuthorizationServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+        PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantId(MultitenantConstants.SUPER_TENANT_ID, true);
+    }
+
+    private RegistryService getRegistryService() throws RegistryException, UserStoreException {
+        RealmService realmService = new InMemoryRealmService();
+        PrivilegedCarbonContext.getThreadLocalCarbonContext().setUsername(ADMIN_USER);
+        BasicDataSource dataSource = new BasicDataSource();
+        String connectionUrl = "jdbc:h2:./target/databasetest/CARBON_TEST";
+        dataSource.setUrl(connectionUrl);
+        dataSource.setDriverClassName("org.h2.Driver");
```
Comment (pamod) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/test/java/org/wso2/carbon/device/mgt/core/authorization/DeviceAccessAuthorizationServiceTest.java:137]: Let's define a constant for org.h2.Driver ?

Comment (GDLMadushanka) [reply components/device-mgt/org.wso2.carbon.device.mgt.core/src/test/java/org/wso2/carbon/device/mgt/core/authorization/DeviceAccessAuthorizationServiceTest.java:137]: +1 fixed in 9ef622da4d64b16865314ac54c0af3d1e208cbcc

**Pull request merged**